src/base/memory_monitor_sr.hpp

```diff
@@ -0,0 +1,271 @@
+#include <type_traits>
+#include <cstdint>
+#include <atomic>
+#include <vector>
+#include <cassert>
+
+#include "perf_def.hpp"
+#include "thread_compat.hpp"
+#include "memory_monitor_common.h"
+
+#define MMM_PARSE_MASK 0x0000FFFF
+#define MMM_MAKE_MODULE_INIT_STAT_ID(module_idx) ((module_idx) << 16)
+#define MMM_MODULE_IDX_FROM_STAT_ID(stat) ((stat) >> 16)
+#define MMM_COMP_INFO_IDX_FROM_STAT_ID(stat) ((stat) & MMM_PARSE_MASK)
+
+typedef enum
+{
+  HEAP = MMM_MAKE_MODULE_INIT_STAT_ID (MMM_HEAP_MODULE),
+  HEAP_ATTRINFO,
+  MMM_STAT_LAST = MMM_MAKE_MODULE_INIT_STAT_ID (MMM_MODULE_LAST)
+} MMM_STAT_ID;
+
+int memmon_add_stat (THREAD_ENTRY *thread_p, MMM_STAT_ID stat_id, uint64_t size);
+int memmon_sub_stat (THREAD_ENTRY *thread_p, MMM_STAT_ID stat_id, uint64_t size);
+int memmon_move_stat (THREAD_ENTRY *thread_p, MMM_STAT_ID src, MMM_STAT_ID dest, uint64_t size);
+int memmon_resize_stat (THREAD_ENTRY *thread_p, MMM_STAT_ID stat_id, uint64_t old_size, uint64_t new_size);
+
+namespace cubperf
+{
+  class memory_monitoring_manager;
+  class mmm_aggregater;
+
+  typedef struct mmm_comp_info
+  {
+    MMM_STAT_ID idx;
+    const char *comp_name;
+    const char *subcomp_name;
+  } MMM_COMP_INFO;
+
+  typedef struct mmm_mem_stat
+  {
+    std::atomic<uint64_t> init_stat;
+    std::atomic<uint64_t> cur_stat;
+    std::atomic<uint64_t> peak_stat;
+    std::atomic<uint32_t> expand_count;
+  } MMM_MEM_STAT;
+
+  MMM_COMP_INFO heap_comp_info[] =
+  {
+    {HEAP, "", ""},
+    {HEAP_ATTRINFO, "ATTRINFO CACHE", ""}
+  };
+
+  class mmm_subcomponent
+  {
+    public:
+      mmm_subcomponent (const char *name)
+      {
+	m_subcomp_name = new char[strlen (name) + 1];
+	strcpy (m_subcomp_name, name);
+      }
+
+      /* public variable */
+      std::atomic<uint64_t> m_cur_stat;
+
+      /* function */
+      char *get_name ()
+      {
+	return m_subcomp_name;
+      }
+
+      ~mmm_subcomponent ()
+      {
+	delete[] m_subcomp_name;
+      }
+    private:
+      /* private variable */
+      char *m_subcomp_name;
+  };
+
+  class mmm_component
+  {
+    public:
+      mmm_component (const char *name)
+      {
+	m_comp_name = new char[strlen (name) + 1];
+	strcpy (m_comp_name, name);
+      }
+
+      /* public variable */
+      MMM_MEM_STAT m_stat;
+      std::vector<mmm_subcomponent> m_subcomponent;
+
+      /* function */
+      int add_subcomponent (const char *name);
+      char *get_name ()
+      {
+	return m_comp_name;
+      }
+
+      ~mmm_component ()
+      {
+	delete[] m_comp_name;
+      }
+    private:
+      /* private variable */
+      char *m_comp_name;
+  };
+
+  class mmm_module
+  {
+    public:
+      mmm_module (const char *name, MMM_COMP_INFO *info);
+      mmm_module () {} /* for dummy */
+
+      /* const */
+      /* max index of component or subcomponent is 0x0000FFFF
+       * if some stats have max index of component or subcomponent,
+       * we don't have to increase it */
+      static constexpr int m_max_idx = 0x0000FFFF;
```

Comment:
It's not a member variable.
```suggestion
      static constexpr int MAX_COMP_IDX = 0x0000FFFF;
```